### ISSUE TITLE
Regression with inline "function new inlining"

### ISSUE THREAD

**Issue by delahee** [HaxeFoundation/haxe#4575]
@ncannasse @Simn 

Hi guys, 

I made the PR test (https://github.com/HaxeFoundation/haxe/pull/4558) and stlil waiting for resolution.
The problem induced by this allocation is very serious. It must absolutely be fixed as it slowsdown very much particle calculus and dynamic geometry creation.

Thanks a lot to make a follow up, please.

**Comment by Simn** [comment]
Is this really a regression?

**Comment by delahee** [comment]
@simn You have the eye of the tiger :D I spent some time to examine the level of regression.
You are partly right, though... 

source :
```
var v = new Inl(5);
var vRand = new Inl(Math.random());
```

Tested on haxe 3.1.3 April 2014

js:
```
var v_x = 5;
var vRand = new Inl(Math.random());
```

cpp:
```
Float v_x = (int)5;		
Float _g7 = ::Math_obj::random();		
Float vRand_x = _g7;		
```

Tested on haxe dev from one week old : 

js:
```
var v_x = 5;
var vRand = new Inl(Math.random());
```

cpp:
```
Float v_x = (int)5;
Float tmp28 = ::Math_obj::random();
::Inl tmp29 = ::Inl_obj::__new(tmp28);
::Inl vRand = tmp29;
```

To sum up : 
Js behaviour was wrong and still is. That is no regression.
cpp behaviour has changed and was way better before. That is a regression.

so yeah it is ahem... in place of one constant and one call, we now have one allocation and a stack copy.

It might seem a little thing to app dev but to me and other gamedev it is a very big issue.

Thanks.
